File: packages/types/src/api/response.ts
import { z } from "zod";
import { ErrorCodeSchema } from "./error-codes";

export const ValidationErrorSchema = z.object({
  code: z.string(),
  field: z.string().optional(),
  message: z.string(),
});

export type ValidationError = z.infer<typeof ValidationErrorSchema>;

export const createApiResponseSchema = <T extends z.ZodType>(
  dataSchema?: T,
) => {
  const baseShape = {
    success: z.boolean(),
    message: z.string(),
    code: ErrorCodeSchema.optional(),
    errors: z.array(ValidationErrorSchema).optional(),
    meta: z.unknown().optional(),
  };

  if (dataSchema) {
    return z.object({
      ...baseShape,
      data: dataSchema,
    });
  }

  return z.object(baseShape);
};

export const apiErrorResponseSchema = z.object({
  success: z.literal(false),
  message: z.string(),
  code: z.string(),
  errors: z.array(ValidationErrorSchema).optional(),
});
export type ApiErrorResponse = z.infer<typeof apiErrorResponseSchema>;

export type ApiResponse<T = unknown> = {
  success: boolean;
  message: string;
  code?: z.infer<typeof ErrorCodeSchema>;
  errors?: ValidationError[];
  data?: T;
  meta?: unknown;
};

const paginationMetaSchema = z.object({
  total: z.number(),
  page: z.number(),
  limit: z.number(),
  total_pages: z.number(),
  offset: z.number(),
});

export type PaginationMeta = z.infer<typeof paginationMetaSchema>;

export type ApiSuccessResponse<T = unknown> = {
  success: true;
  message: string;
  data?: T;
  meta?: PaginationMeta;
};

File: apps/api/internal/types/response.go
package types

type AppError struct {
	Code    string `json:"code"`
	Field   string `json:"field,omitempty"`
	Message string `json:"message"`
}

type APIResponse struct {
	Success bool       `json:"success"`
	Message string     `json:"message,omitempty"`
	Code    string     `json:"code,omitempty"`
	Errors  []AppError `json:"errors,omitempty"`
	Data    any        `json:"data,omitempty"`
	Meta    any        `json:"meta,omitempty"`
}

type PaginationMeta struct {
	Total      int `json:"total"`
	TotalPages int `json:"total_pages"`
	Page       int `json:"page"`
	Limit      int `json:"limit"`
	Offset     int `json:"offset"`
}
